dask/array/tests/test_creation.py

```diff
@@ -89,6 +92,51 @@ def test_linspace(endpoint):
             sorted(da.linspace(6, 49, endpoint=endpoint, chunks=5, dtype=float).dask))
 
 
+def _from_block_function_callback(block_info, calls, lock):
+    """Callback used in tests of from_block_function.
+
+    It generates a 2D array where each cell has value 1000*row+col.
+    It also appends the received `block_info` to `calls`, while
+    holding `lock`.
+    """
+    loc = block_info.array_location
+    rows = np.arange(loc[0][0], loc[0][1])[:, np.newaxis]
+    cols = np.arange(loc[1][0], loc[1][1])[np.newaxis, :]
+    with lock:
+        calls.append(block_info)
+    return 1000 * rows + cols
+
+
+@pytest.mark.parametrize('dtype', [None, np.int_])
+def test_from_block_function(dtype):
+    shape = (3, 4)
+    chunks = ((1, 2), 2)
+    num_chunks = (2, 2)
+    calls = []
+    lock = threading.Lock()
+    # Passes calls via *args and lock via **kwargs to test both
+    a = da.from_block_function(_from_block_function_callback, calls,
+                               shape=shape, chunks=chunks, dtype=dtype,
+                               lock=lock)
+    expected = np.array([[0, 1, 2, 3], [1000, 1001, 1002, 1003], [2000, 2001, 2002, 2003]])
+    assert_eq(a, expected)
+
+    if dtype is None:
+        # First call will be just to determine dtype
+        del calls[0]
+    calls.sort(key=lambda call: call.chunk_location)
+    assert calls == [
+        FromBlockFunctionInfo(shape=shape, num_chunks=num_chunks, chunk_location=(0, 0),
+                              array_location=[(0, 1), (0, 2)], chunk_shape=(1, 2), dtype=np.dtype(np.int_)),
+        FromBlockFunctionInfo(shape=shape, num_chunks=num_chunks, chunk_location=(0, 1),
+                              array_location=[(0, 1), (2, 4)], chunk_shape=(1, 2), dtype=np.dtype(np.int_)),
+        FromBlockFunctionInfo(shape=shape, num_chunks=num_chunks, chunk_location=(1, 0),
+                              array_location=[(1, 3), (0, 2)], chunk_shape=(2, 2), dtype=np.dtype(np.int_)),
+        FromBlockFunctionInfo(shape=shape, num_chunks=num_chunks, chunk_location=(1, 1),
+                              array_location=[(1, 3), (2, 4)], chunk_shape=(2, 2), dtype=np.dtype(np.int_))
+    ]
```

Comment:
This test seems somewhat brittle.  If the implementation of `FromBlockFunctionInfo` were to be changed in any way, my guess is that this test might break.  Instead it might be useful to test things about the output array type, rather than internal details of the algorithm.